### PR TITLE
Fixes for issues #24, #25

### PR DESCRIPTION
**Pull request by aluuu**
I've spotted and fixed problem with PING example in #25 (see 3b931af for further details).

cc @agarwal, @0xffea

### CODE REVIEW

**Comment by agarwal** [comment]
Thanks. I'll test it soon.

**Comment by aluuu** [comment]
Moreover, looks like this pull request makes CI happy.

File: test/test.ml [hunk 1]
```diff
@@ -396,32 +391,35 @@ module Make(IO : Redis.S.IO) = struct
...
 
+  let cleanup_keys conn =
+    Client.keys conn "ounit_*" >>= Client.del conn
 
   let bracket test_case () =
-    IO.run
-      (setup () >>= fun conn ->
-       test_case conn >>= fun () ->
-       teardown conn)
+    IO.run @@ Client.with_connection redis_spec test_case
```
Comment (hcarty) [inline test/test.ml]: If the code uses `@@` the OCaml version should be restricted to >= 4.01.0 in `_oasis`.  That should also simplify the code generated by oasis.

File: test/test.ml [hunk 2]
```diff
@@ -396,32 +391,35 @@ module Make(IO : Redis.S.IO) = struct
...
+
+  let teardown () =
+    IO.run @@ Client.with_connection redis_spec cleanup_keys
```
Comment (hcarty) [inline test/test.ml]: Same as above regarding `@@`

File: test/test.ml (hunk 1 above)
Comment (aluuu) [inline test/test.ml]: Hm, `setup.ml` became even longer after restriction. Anyway, thanks for review.

Comment (hcarty) [inline test/test.ml]: I didn't expect that... but it should be helpful none the less, particularly when this makes it into opam.

**Comment by aluuu** [comment]
I've met a problem with `_tags` file generation by OASIS after adding `OCamlVersion` restriction.

Steps to reproduce the problem:
```
> opam remove lwt
> cd <ocaml-redis-path>
> ocaml setup.ml -clean
> ocaml setup.ml -configure --disable-lwt
> ocaml setup.ml -build

+ /Users/admin/.opam/4.03.0/bin/ocamlfind ocamldep -package uuidm -package re.str -package lwt.unix -package lwt -modules src/redis_sync.mli > src/redis_sync.mli.depends
ocamlfind: Package `lwt.unix' not found
Command exited with code 2.
``` 

I assume, that flags `-package lwt.unix -package lwt` should not be in `ocamlfind` invocation.
As I see, OASIS changed all `pkg_*` tags to `package(*)` tags (30d33e4) and it requires `package(lwt)` for all source files in `src/` directory, which is not correct (as far as `redis_sync` doesn't use lwt).

For fix this problem I have two options:
1. Manually fix `_tags` file to use lwt package only for `redis_lwt.ml(i)` files and disable `_tags` file auto-generation by OASIS.
2. Get rid of OASIS.

Is there a way to disable auto-generation of `_tags` file by OASIS?

**Comment by aluuu** [comment]
@agarwal ping. Pull request needs your feedback before merge :)

**Comment by Drup** [comment]
A solution that doesn't involve putting stuff in `_tags`: move the redis_lwt library in a different directory than `src/`. oasis applies tags directory per directory, so the problem is going to appear for each tag, not only `package`.

**Comment by agarwal** [comment]
Sorry, I can't confirm. Just tried for about 1 hour, but I keep hitting various issues. First I still hit #21, so I had to try and merge that fix in with this pull request. However, that turns out to be no so straightforward because I keep getting merge conflicts between oasis's auto-generated files.

Also, I suggest following GitHub's recommendation on pull requests, which is to create a new branch in your fork for each pull request. That way each branch is isolated in terms of what it is trying to fix. Since I'm hitting #21, that might not have helped me in this case, but maybe my merge conflicts would have been easier to resolve.

**Comment by aluuu** [comment]
Ok, then I merge this PR, because it fixes issues #24 and #25. 
In the next pull request I'll try to rework project structure to avoid tricky `_tags` additions and your `Mutex` module problem.

**Comment by agarwal** [comment]
Thanks for the work on this.

The auto-generated files of oasis is a longstanding annoyance. I guess the `-setup-update` flag was supposed to resolve this problem, but I haven't used oasis in years so not sure anymore.

**Pull request merged**